designs/disruption-controls-by-action.md

```diff
@@ -0,0 +1,188 @@
+1. User needs to be able to define an action and a budget/budgets for that action. 
+2. Actions that should be supported should by any disruption actions effected by the current Budgets implementation. (Consolidation, Emptiness, Expiration, Drift) 
+3. Budgets should still support the option to set a given behavior for all disruption actions. If action is unspecified we want to just assume its mean't to apply to all actions. 
+
+## API Design
+### Approach A: Add an action field to disruption Budgets 
+This approach outlines a simple api change to the betav1 nodepool api to allow disruption budgets to specify a disruption action. 
+### Proposed Spec
+Add a simple field "action" is proposed to be added to the budgets. 
+```go
+// Budget defines when Karpenter will restrict the
+// number of Node Claims that can be terminating simultaneously.
+type Budget struct {
+	// Action defines the disruption action that this particular disruption budget applies to. 
+	Action DisruptionAction `json:"action,omitempty" hash:"ignore"`
+	// Nodes dictates the maximum number of NodeClaims owned by this NodePool
+	// that can be terminating at once. This is calculated by counting nodes that
+	// have a deletion timestamp set, or are actively being deleted by Karpenter.
+	// This field is required when specifying a budget.
+	// This cannot be of type intstr.IntOrString since kubebuilder doesn't support pattern
+	// checking for int nodes for IntOrString nodes.
+	// Ref: https://github.com/kubernetes-sigs/controller-tools/blob/55efe4be40394a288216dab63156b0a64fb82929/pkg/crd/markers/validation.go#L379-L388
+	// +kubebuilder:validation:Pattern:="^((100|[0-9]{1,2})%|[0-9]+)$"
+	// +kubebuilder:default:="10%"
+	Nodes string `json:"nodes" hash:"ignore"`
+	// Schedule specifies when a budget begins being active, following
+	// the upstream cronjob syntax. If omitted, the budget is always active.
+	// Timezones are not supported.
+	// This field is required if Duration is set.
+	// +kubebuilder:validation:Pattern:=`^(@(annually|yearly|monthly|weekly|daily|midnight|hourly))|((.+)\s(.+)\s(.+)\s(.+)\s(.+))$`
+	// +optional
+	Schedule *string `json:"schedule,omitempty" hash:"ignore"`
+	// Duration determines how long a Budget is active since each Schedule hit.
+	// Only minutes and hours are accepted, as cron does not work in seconds.
+	// If omitted, the budget is always active.
+	// This is required if Schedule is set.
+	// This regex has an optional 0s at the end since the duration.String() always adds
+	// a 0s at the end.
+	// +kubebuilder:validation:Pattern=`^([0-9]+(m|h)+(0s)?)$`
+	// +kubebuilder:validation:Type="string"
+	// +optional
+	Duration *metav1.Duration `json:"duration,omitempty" hash:"ignore"`
+}
+
+type DisruptionAction string 
+
+const (
+	All DisruptionAction = "All"
+	Consolidation DisruptionAction = "Consolidation" 
+	Drift DisruptionAction = "Drift" 
+	Expiration DisruptionAction = "Expiration" 
+	Emptiness DisruptionAction = "Emptiness"
```

Comment:
@jonathan-innis  thoughts on also adding more granularity to drift and being able to specify how many at a time should be disrupted for given drift reasons? FOr Azure provider we have K8sVersionDrift, and NodeImageDrift, then there is actual drift. I am wondering if users may want to set different windows for Upgrade vs drift proper.